internal/service/scheduler/snapshot/operator_profit/operator_profit.go

```diff
@@ -99,59 +101,62 @@ func (s *server) Run(ctx context.Context) error {
 	return nil
 }
 
-func (s *server) saveOperatorProfitSnapshots(ctx context.Context, latestEpochSnapshot, latestEpochEvent uint64) error {
+// beginSnapshot takes new snapshots of all Nodes' operation pool up to the latest epoch.
+func (s *server) beginSnapshot(ctx context.Context, currentSnapshotEpochID, latestEpochID uint64) error {
 	// Query the array of Nodes.
-	nodes, err := s.databaseClient.FindNodes(ctx, schema.FindNodesQuery{})
+	nodes, err := s.getNodesFromDB(ctx)
 	if err != nil {
-		zap.L().Error("find Nodes", zap.Error(err))
-
-		return fmt.Errorf("find Nodes: %w", err)
+		return fmt.Errorf("get nodes: %w", err)
 	}
 
-	for epochID := latestEpochSnapshot + 1; epochID <= latestEpochEvent; epochID++ {
-		// Fetch the epoch items by the epoch id.
-		epochItems, err := s.databaseClient.FindEpochTransactions(ctx, epochID, 1, nil)
+	// Iterate until the snapshot is up to date with the latest epoch.
+	// currentEpochID is the epoch being snapshotted.
+	for currentEpochID := currentSnapshotEpochID + 1; currentEpochID <= latestEpochID; currentEpochID++ {
+		// Fetch the epoch by currentEpochID.
+		epoch, err := s.databaseClient.FindEpochTransactions(ctx, currentEpochID, 1, nil)
 		if err != nil {
 			zap.L().Error("find epoch transactions", zap.Error(err))
 
 			continue
 		}
 
-		if len(epochItems) == 0 {
+		// If the epoch does not exist in the database, log an error and continue.
+		// This means epochs are not indexed up to the latest epoch.
+		if len(epoch) == 0 {
+			zap.L().Error("an epoch does not exist in database", zap.Any("epoch ID", currentEpochID))
+
+			continue
+		}
+
+		// If there are no nodes, continue to the next epoch.
+		if len(nodes) == 0 {
 			continue
 		}
 
 		var (
-			mutex     sync.Mutex
 			errorPool = pool.New().WithContext(ctx).WithMaxGoroutines(30).WithCancelOnError().WithFirstError()
-			data      = make([]*schema.OperatorProfitSnapshot, 0, len(nodes))
+			result    = make([]*schema.OperationPoolSnapshot, len(nodes))
```

Comment:
Based on the code, it should be named "OperationPoolSnapshot" and that is what this struct and the whole cron job is doing. It has nothing (and should not have) to do with "OperatorProfit"